File: src/QsCompiler/Core/Transformations/TypeTransformation.fs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

namespace Microsoft.Quantum.QsCompiler.Transformations.Core

open System.Collections.Immutable
open Microsoft.Quantum.QsCompiler.DataTypes
open Microsoft.Quantum.QsCompiler.SyntaxExtensions
open Microsoft.Quantum.QsCompiler.SyntaxTokens
open Microsoft.Quantum.QsCompiler.SyntaxTree
open Microsoft.Quantum.QsCompiler.Transformations.Core.Utils

type private ExpressionType = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>


type TypeTransformationBase(options: TransformationOptions) =

    let Node = if options.Rebuild then Fold else Walk
    new() = new TypeTransformationBase(TransformationOptions.Default)


    // supplementary type information

    abstract OnRangeInformation: QsNullable<Range> -> QsNullable<Range>
    default this.OnRangeInformation range = range

    abstract OnCharacteristicsExpression: ResolvedCharacteristics -> ResolvedCharacteristics
    default this.OnCharacteristicsExpression fs = fs

    abstract OnCallableInformation: CallableInformation -> CallableInformation

    default this.OnCallableInformation opInfo =
        let characteristics =
            this.OnCharacteristicsExpression opInfo.Characteristics

        let inferred = opInfo.InferredInformation

        CallableInformation.New |> Node.BuildOr opInfo (characteristics, inferred)


    // nodes containing subtypes

    abstract OnUserDefinedType: UserDefinedType -> ExpressionType

    default this.OnUserDefinedType udt =
        let ns, name = udt.Namespace, udt.Name
        let range = this.OnRangeInformation udt.Range

        ExpressionType.UserDefinedType << UserDefinedType.New |> Node.BuildOr InvalidType (ns, name, range)

    abstract OnTypeParameter: QsTypeParameter -> ExpressionType

    default this.OnTypeParameter tp =
        let origin = tp.Origin
        let name = tp.TypeName
        let range = this.OnRangeInformation tp.Range

        ExpressionType.TypeParameter << QsTypeParameter.New
        |> Node.BuildOr InvalidType (origin, name, range)

    abstract OnOperation: (ResolvedType * ResolvedType) * CallableInformation -> ExpressionType

    default this.OnOperation((it, ot), info) =
        let transformed =
            (this.OnType it, this.OnType ot), this.OnCallableInformation info

        ExpressionType.Operation |> Node.BuildOr InvalidType transformed

    abstract OnFunction: ResolvedType * ResolvedType -> ExpressionType

    default this.OnFunction(it, ot) =
        let transformed = this.OnType it, this.OnType ot

        ExpressionType.Function |> Node.BuildOr InvalidType transformed

    abstract OnTupleType: ImmutableArray<ResolvedType> -> ExpressionType

    default this.OnTupleType ts =
        let transformed =
            ts |> Seq.map this.OnType |> ImmutableArray.CreateRange

        ExpressionType.TupleType |> Node.BuildOr InvalidType transformed

    abstract OnArrayType: ResolvedType -> ExpressionType

    default this.OnArrayType b =
        ExpressionType.ArrayType |> Node.BuildOr InvalidType (this.OnType b)


    // leaf nodes

    abstract OnUnitType: unit -> ExpressionType
    default this.OnUnitType() = ExpressionType.UnitType

    abstract OnQubit: unit -> ExpressionType
    default this.OnQubit() = ExpressionType.Qubit

    abstract OnMissingType: unit -> ExpressionType
    default this.OnMissingType() = ExpressionType.MissingType

    abstract OnInvalidType: unit -> ExpressionType
    default this.OnInvalidType() = ExpressionType.InvalidType

    abstract OnInt: unit -> ExpressionType
    default this.OnInt() = ExpressionType.Int

    abstract OnBigInt: unit -> ExpressionType
    default this.OnBigInt() = ExpressionType.BigInt

    abstract OnDouble: unit -> ExpressionType
    default this.OnDouble() = ExpressionType.Double

    abstract OnBool: unit -> ExpressionType
    default this.OnBool() = ExpressionType.Bool

    abstract OnString: unit -> ExpressionType
    default this.OnString() = ExpressionType.String

    abstract OnResult: unit -> ExpressionType
    default this.OnResult() = ExpressionType.Result

    abstract OnPauli: unit -> ExpressionType
    default this.OnPauli() = ExpressionType.Pauli

    abstract OnRange: unit -> ExpressionType
    default this.OnRange() = ExpressionType.Range


    // transformation root called on each node

    member this.OnType(t: ResolvedType) =
        if not options.Enable then
            t
        else
            let transformed =
                t.Resolution
                |> function
                | ExpressionType.UnitType -> this.OnUnitType()
                | ExpressionType.Operation ((it, ot), fs) -> this.OnOperation((it, ot), fs)
                | ExpressionType.Function (it, ot) -> this.OnFunction(it, ot)
                | ExpressionType.TupleType ts -> this.OnTupleType ts
                | ExpressionType.ArrayType b -> this.OnArrayType b
                | ExpressionType.UserDefinedType udt -> this.OnUserDefinedType udt
                | ExpressionType.TypeParameter tp -> this.OnTypeParameter tp
                | ExpressionType.Qubit -> this.OnQubit()
                | ExpressionType.MissingType -> this.OnMissingType()
                | ExpressionType.InvalidType -> this.OnInvalidType()
                | ExpressionType.Int -> this.OnInt()
                | ExpressionType.BigInt -> this.OnBigInt()
                | ExpressionType.Double -> this.OnDouble()
                | ExpressionType.Bool -> this.OnBool()
                | ExpressionType.String -> this.OnString()
                | ExpressionType.Result -> this.OnResult()
                | ExpressionType.Pauli -> this.OnPauli()
                | ExpressionType.Range -> this.OnRange()

            let ResolvedType t = ResolvedType.New(true, t)
            ResolvedType |> Node.BuildOr t transformed

File: src/QsCompiler/QirGeneration/Subtransformations/TypeTransformation.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using LlvmBindings.Types;
using Microsoft.Quantum.QIR;
using Microsoft.Quantum.QsCompiler.SyntaxTokens;
using Microsoft.Quantum.QsCompiler.SyntaxTree;
using Microsoft.Quantum.QsCompiler.Transformations.Core;
using Microsoft.Quantum.QsCompiler.Transformations.QsCodeOutput;

namespace Microsoft.Quantum.QsCompiler.QIR
{
    using QsResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;

    internal class QirTypeTransformation
    {
        private readonly QirGlobalType globalType;

        public QirTypeTransformation(Types types, Func<QsQualifiedName, QsCustomType?> getTypeDecl) =>
            this.globalType = new QirGlobalType(types, getTypeDecl);

        internal ITypeRef LlvmTypeFromQsharpType(ResolvedType resolvedType) =>
            this.globalType.LlvmTypeFromQsharpType(resolvedType);

        private class QirGlobalType : TypeTransformation
        {
            private protected ITypeRef? BuiltType { get; set; }

            private protected Types QirTypes { get; }

            private protected Func<QsQualifiedName, QsCustomType?> TypeDeclaration { get; }

            public QirGlobalType(Types types, Func<QsQualifiedName, QsCustomType?> getTypeDecl)
            : base(TransformationOptions.NoRebuild)
            {
                this.QirTypes = types;
                this.TypeDeclaration = getTypeDecl;
            }

            /// <summary>
            /// Gets the QIR equivalent for a Q# type.
            /// Tuples are represented as QirTuplePointer, arrays as QirArray, and callables as QirCallable.
            /// </summary>
            /// <param name="resolvedType">The Q# type</param>
            /// <returns>The equivalent QIR type</returns>
            internal ITypeRef LlvmTypeFromQsharpType(ResolvedType resolvedType)
            {
                this.BuiltType = null;
                this.OnType(resolvedType);
                return this.BuiltType ?? throw new NotImplementedException(
                    $"Llvm type for {SyntaxTreeToQsharp.Default.ToCode(resolvedType)} could not be constructed.");
            }

            /// <summary>
            /// Creates a pointer to the concrete type of a QIR tuple value that contains items of the given types.
            /// </summary>
            private protected virtual ITypeRef CreateConcreteTupleType(IEnumerable<ResolvedType> items) =>
                this.QirTypes.TypedTuple(items.Select(this.LlvmTypeFromQsharpType)).CreatePointerType();

            /* public overrides */

            public override QsResolvedTypeKind OnArrayType(ResolvedType b)
            {
                this.BuiltType = this.QirTypes.Array;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnBigInt()
            {
                this.BuiltType = this.QirTypes.BigInt;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnBool()
            {
                this.BuiltType = this.QirTypes.Bool;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnDouble()
            {
                this.BuiltType = this.QirTypes.Double;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnFunction(ResolvedType it, ResolvedType ot)
            {
                this.BuiltType = this.QirTypes.Callable;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnInt()
            {
                this.BuiltType = this.QirTypes.Int;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnOperation(Tuple<ResolvedType, ResolvedType> _arg1, CallableInformation info)
            {
                this.BuiltType = this.QirTypes.Callable;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnPauli()
            {
                this.BuiltType = this.QirTypes.Pauli;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnQubit()
            {
                this.BuiltType = this.QirTypes.Qubit;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnRange()
            {
                this.BuiltType = this.QirTypes.Range;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnResult()
            {
                this.BuiltType = this.QirTypes.Result;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnString()
            {
                this.BuiltType = this.QirTypes.String;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnTupleType(ImmutableArray<ResolvedType> ts)
            {
                this.BuiltType = this.CreateConcreteTupleType(ts);
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnUnitType()
            {
                // Unit is represented as a null tuple pointer (an empty tuple).
                // This is necessary because "void" in LLVM is not a proper type and can't be included
                // as an element in a struct.
                this.BuiltType = this.QirTypes.Tuple;
                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnUserDefinedType(UserDefinedType udt)
            {
                // User-defined types are represented by a tuple of their items.
                var udtDefinition = this.TypeDeclaration(udt.GetFullName());
                if (udtDefinition != null)
                {
                    var resolvedType = udtDefinition.Type;
                    this.BuiltType = resolvedType.Resolution is QsResolvedTypeKind.TupleType tuple
                        ? this.CreateConcreteTupleType(tuple.Item)
                        : this.CreateConcreteTupleType(new[] { resolvedType });
                }
                else
                {
                    throw new InvalidOperationException("unknown user defined type");
                }

                return QsResolvedTypeKind.InvalidType;
            }

            public override QsResolvedTypeKind OnTypeParameter(QsTypeParameter tp)
            {
                this.BuiltType = this.QirTypes.BytePointer;
                return QsResolvedTypeKind.InvalidType;
            }
        }
    }
}
